### DISCUSSION TITLE
Php empty function returns true on Livewire computed property, help with multiple <select> fields.

### DISCUSSION THREAD

**Discussion by modrictin** [livewire/livewire#4366] [7. Bugs (archived)] [answered]
Hi all,

I have a problem with computed properties.
I defined Test computed property as shown below:
`
 public function getTestProperty(){
        return [1,3,2];
    }
`

When I'm in my view, if i do ` @php dd($this->test); @endphp ` I get an array [1,3,2] as expected.

If I do ` @php dd(empty($this->test)); @endphp ` I get true...
Why is this happening?

-----------

One more unrelated question:

I have 4 `<select>` fields whose data change depending on the previous selection.
Is it possible to select the first value in `<select>` field when data changes and bind it to the model without doing it programmatically and defining 4 watchers/hooks who call themself to reset data?
For example, when I change `<select>` number 1 I need to reset select numbers 2, 3, and 4. And when I change `<select>` 2 I need to reset `<select>` 3 and 4. All of this is giving me a  spaghetti code because in every hook/watcher I need to get and reset/set the data for the next 3/2/1 select fields/models or the data in the model stays from the old selected data.

Is there some other way to approach this?

**Comment by joshhanley** [comment, marked as answer]
@progonitelj the first issue is an issue with PHP. Using `empty()` won't recognise the computed property, so you should try using `isset()` instead. As to why, well welcome to the wonders of PHP, sometimes it has some obscure ways that it works! 😕

In regards to your second question, please make a seperate discussion in future. But yeah updated hooks are probably your best option, but it shouldn't end up like spaghetti code. Can you share your component class with that setup?

**Reply by modrictin** [reply]
```
class AddNewReservation extends Component
{
    public $reservation;
    public $tableOccupancy;
    public $transferData;
    public $availableTerraceTables;
    public $terraceId;
    public $tables;
    public $customTime = false;
    public  $tableReservations=[];

    public $slots = [];
    protected $rules = [
        'reservation.num_people' => 'required|numeric',
        'reservation.guest_full_name' => 'required|min:3',
        'reservation.important' => '',
        'reservation.tasting_menu' => '',
        'reservation.deposit' => '',
        'reservation.transfer' => '',
        'reservation.notice' => '',
        'reservation.phone' => '',
        'reservation.email' => 'required',
        'reservation.exclusive' => '',

        'tableOccupancy.terrace_slot_id' => '',
        'tableOccupancy.date' => 'required|date',
        'tableOccupancy.time_from' => 'required',
        'tableOccupancy.time_to' => 'required',
        'tableOccupancy.table_id' => 'required',
    ];

    public function mount()
    {
        $this->reservation = new TableReservation();
        $this->tableOccupancy = new tableOccupancy();
        $this->reservation->num_people = 1;
        $this->setLivewireSelectData();
    }

    #Livewire wont set the model value as first item in <select> so we have to do it manually
    private function setLivewireSelectData()
    {
        $this->reservation->transfer = 1;
        $this->terraceId = 1;
        $this->tableOccupancy->date = Carbon::now()->format('Y-m-d');
        $this->updatedTerraceId();
    }

    //Watchers + live validation
    public function updated($propertyName)
    {
        if ($propertyName == 'tableOccupancy.table_id') {
            $this->updatedTableOccupancyTableId();
        }
        if ($propertyName == 'tableOccupancy.terrace_slot_id') {
            $this->updatedTableOccupancyTableId();
        }
        if ($propertyName == 'tableOccupancy.date') {
            $this->updatedTableOccupancyDate();
        }
        if (in_array($propertyName, ['tableOccupancy.time_from', 'tableOccupancy.time_to'])) {
            $this->updatedTimeData();
        }
        //Live validation
        $this->validateOnly($propertyName);
    }

    #We dont need to put this in updated() method because Livewire will trigger it automaticly because it's not multiword AND nested property :)
    public function updatedTerraceId()
    {
        $this->tables = Table::whereTerraceId($this->terraceId)->get();
        $this->tableOccupancy->table_id = $this->tables->first()->id;

        $this->slots = [];
        $slotDateRange =
            SlotDateRange::where('date_from', '<=', $this->tableOccupancy->date)
                ->where('date_to', '>=', $this->tableOccupancy->date)
                ->where('terrace_id', '=', $this->terraceId)
                ->with('slots:id,slot_date_range_id,time_from,time_to')->first();

        if ($slotDateRange)
            $this->slots = $slotDateRange->slots->keyBy('id')->toArray();

        if (!empty($this->slots)) {
            //set data for the changed <select> and update time from and time to in the tablle occupancy model
            $this->tableOccupancy->terrace_slot_id = $slotDateRange->slots->first()->id;
            $this->customTime = false;
        } else {
            $this->customTime = true;
            $this->tableOccupancy->terrace_slot_id = null;
        }
        $this->updatedTableOccupancyTerraceSlotId();

        $this->updatedCustomTime();

    }

    public function updatedTableOccupancyTerraceSlotId()
    {
        if (!empty($this->slots) && isset($this->slots[$this->tableOccupancy->terrace_slot_id])) {
            $this->tableOccupancy->time_from = $this->slots[$this->tableOccupancy->terrace_slot_id]['time_from'];
            $this->tableOccupancy->time_to = $this->slots[$this->tableOccupancy->terrace_slot_id]['time_to'];
        }
        $this->validateReservationLogic();
    }

    public function updatedCustomTime()
    {
        if ($this->customTime == true) {
            $this->tableOccupancy->terrace_slot_id = null;
        } else {
            $this->tableOccupancy->terrace_slot_id = Arr::first($this->slots)['id'];
        }
        $this->validateReservationLogic();

    }

    private function updatedTimeData()
    {
        if ($this->tableOccupancy->time_from > $this->tableOccupancy->time_to) {
            $this->tableOccupancy->time_to = $this->tableOccupancy->time_from;
        }
        $this->validateReservationLogic();

    }

    public function updatedTableOccupancyTableId()
    {
        $this->validateReservationLogic();
    }

    public function updatedTableOccupancyDate()
    {
        $this->updatedTerraceId();
    }



    public function validateReservationLogic()
    {
        $validationStatus = true;
        $this->clearValidation('tableUnavailable');
        $tableOccupancy = TableOccupancy::getTableAvailabilityAndReservations($this->tableOccupancy->date, $this->tableOccupancy->time_from, $this->tableOccupancy->time_to, $this->tableOccupancy->table_id);
        if ($tableOccupancy['status'] == TableOccupancy::STATUS_UNAVAILABLE) {
            $this->addError('tableUnavailable', 'Stol je zauzet! Rezervacije ovog stola za ovo vrijeme su:');
            $validationStatus = false;
            $this->tableReservations = TableReservation::whereIn('id',$tableOccupancy['reservations'])->with('availabilityData:id,table_reservation_id,date,time_from,time_to')->get(['id','num_people','guest_full_name'])->toArray();
        }else{
            $this->tableReservations = [];
        }
        return $validationStatus;
    }


    public function updatingReservationTransfer($value, $key)
    {
        switch ($value) {
            case 2:
                $this->getTransferData();
        }
    }

    public function getTransferData()
    {
        dd($this->getShuttleData());
    }

    public function getShuttleData($type = '1')
    {
        $shuttleSeason = ShuttleSeason::where('date_from', '<=', $this->reservation->date)
            ->where('date_to', '>=', $this->reservation->date)
            ->with('shuttleDepartures')->first();
        if (!$shuttleSeason)
            return [];
        return $shuttleSeason;
    }

    public function addReservation()
    {
        $this->validate();
        if ($this->validateReservationLogic() == false) {
            $this->showToast('Rezervacija nije spremljena', '', 'error');
            return false;
        }
        if (!empty($this->tableOccupancy->terrace_slot_id)) {
            $slot = TerraceSlot::whereId($this->tableOccupancy->terrace_slot_id)->first();

            $this->tableOccupancy->time_from = $slot->time_from;
            $this->tableOccupancy->time_to = $slot->time_to;
        }

        DB::transaction(function () {
            $this->reservation->save();
            $this->reservation->availabilityData()->save($this->tableOccupancy);
        });
    }


    public function render()
    {
        return view('livewire.add-new-reservation');
    }


}
```

Hi there, as you can see it's a pretty big component and it handles importing reservation data.

I have 3 `<select>` fields in which data changes depending on each other selected values.
I could have implemented the computed properties but I decided to do it in the watchers.
What worries me is that I'm doing it the wrong way. But after your answer, I suppose there is no other way.

As you can see, in the mount() I set all the data for the component and the data for the models of `<select>` fields so their values are not empty. 

I don't want to set the `<option>Please select table</option>` as my first value just to force the user to select it, to avoid that I set the data in the mount() and in every other watcher when the data changes. That's why I'm concerned about the spaghetti code. If I don't set the data in the watchers for `<select>` fields that the old data persists in the model values if the user doesn't select it manualy.

As you can see in the code, I have updated() method which calls my other updated methods, I know Livewire handles this out of the box with helpers but I have multiword nested properties.

Then When my date changes, it calls `updatedTerraceId()` which sets the tables and slots for the other two `<select>` fields. Then I set their default value with `->first()` eloquent method. Then I call other watchers to set the data of each other and chain everything down to the last field.

I want to say thanks for helping me on this!

**Reply by joshhanley** [reply]
@progonitelj yeah generally all looks ok to me, sometimes when we have complicated dependencies like this, it's what we have to do.

My only feedback would be to not call the updated hook methods directly (like calling `$this->updatedTerraceId()`), instead what I would do, is extract the code that is in them into methods with descriptive names. Then you can call those methods directly instead, and you can change your hooks to also call those methods.

That will make it cleaner and easier to read as you can name the method properly that describes what that method is doing (and you're not stuck with the `updated...` naming convention).

Also in your generic `updated` hook, you shouldn't need to check the property and manually call it's specific hook as Livewire does that for you, so you could remove all of that code.

Hope this helps!

**Reply by modrictin** [reply]
Thank you so much for your time!

I already refactored my code pretty much, so I'm now left with the cleaner code.
Thanks again!

**Reply by joshhanley** [reply]
@progonitelj no worries! 🙂
